[WIP] Major refactor (breaking changes)

puckey opened the pull request:
I spent some time looking into improving detect-gpu:
- it was returning incorrect results in several cases, which I was able to improve by changing the way getEntryVersionNumber functions
- implemented a bunch of tests
- slimmed down the footprint by splitting up the data and loading it conditionally (for this reason I had to retire the umd build. It does not support conditional loading)
- precomputing 'entry version numbers' and limiting models to check by above should also cause it to take up less cpu
- switched to gfxbench.com as data source, since it offers fps benchmark data, which solves https://github.com/TimvanScherpenzeel/detect-gpu/issues/12
- moved blacklist check into the data files
- replaced `device-ua` dependency with plain js
- instead of returning a string, getGPUTier now returns a promise which resolves to an object with properties: `tier`, `mobile` (boolean), `type` (`BENCHMARK` | `BLACKLISTED` | `WEBGL_UNSUPPORTED` | `FALLBACK`), and `model` (the renderer string it matched – handy for tests).

Instead of looking at the changed files, it probably makes sense to look at my repo instead: https://github.com/puckey/detect-gpu

Open for feedback on this – I realise it is a lot in one go.

TimvanScherpenzeel commented:
Excellent! Thank you for putting so much effort into this! I'll have a good look over the weekend and get back to you.

TimvanScherpenzeel commented on package.json:
```diff
@@ -16,39 +16,32 @@
     "build": "tsc --module commonjs && rollup -c rollup.config.ts",
     "parse-analytics": "node ./scripts/analytics_parser.js",
-    "update-benchmarks": "node ./scripts/update_benchmarks.js"
+    "update-benchmarks": "rimraf src/data && mkdir src/data && ts-node -O '{\"module\":\"commonjs\"}' ./scripts/update_benchmarks.ts"
```
To be safe `mkdir -p` would be good to use

TimvanScherpenzeel commented on src/internal/device.ts:
```diff
@@ -0,0 +1,18 @@
+
+  return {
+    mobile: isAndroid || isIOS || isIPad,
+    safari12: /Version\/12.+Safari/.test(userAgent)
```
I would prefer the use of `isSafari12` over `safari12`

TimvanScherpenzeel commented on src/internal/device.ts:
```diff
@@ -0,0 +1,18 @@
+  const isAndroid = /android/i.test(userAgent);
+
+  return {
+    mobile: isAndroid || isIOS || isIPad,
```
I think, looking at the combined market share of iOS and Android these tests will suffice.

TimvanScherpenzeel commented on src/index.ts:
```diff
@@ -1,164 +1,121 @@
-            distance: getLevenshteinDistance(renderer, entry),
-          });
-        }
+  mobile: boolean
```
would prefer the `isMobile` syntax

puckey commented:
Thanks!

I am wondering if we shouldn't use the raw `fps` information instead of calculating percentiles, since the percentiles will change whenever a device comes on the scene with higher frame rates.

And I am considering switching to the following benchmark, which seems to allow for higher fpses on lower powered devices: https://gfxbench.com/result.jsp?benchmark=gfx50&test=756&text-filter=&order=median&ff-lmobile=true&ff-smobile=true&os-Android_gl=true&os-Android_vulkan=true&os-iOS_gl=true&os-iOS_metal=true&os-Linux_gl=true&os-OS_X_gl=true&os-OS_X_metal=true&os-Windows_dx=true&os-Windows_dx12=true&os-Windows_gl=true&os-Windows_vulkan=true&pu-dGPU=true&pu-iGPU=true&pu-GPU=true&arch-ARM=true&arch-unknown=true&arch-x86=true&base=device

puckey commented:
@TimvanScherpenzeel (Maybe wait with adding commits to this branch – I am still changing things and don't want to deal with too many merges quite yet.)

TimvanScherpenzeel commented:
> I am wondering if we shouldn't use the raw `fps` information instead of calculating percentiles, since the percentiles will change whenever a device comes on the scene with higher frame rates.

> And I am considering switching to the following benchmark, which seems to allow for higher fpses on lower powered devices: https://gfxbench.com/result.jsp?benchmark=gfx50&test=756&text-filter=&order=median&ff-lmobile=true&ff-smobile=true&os-Android_gl=true&os-Android_vulkan=true&os-iOS_gl=true&os-iOS_metal=true&os-Linux_gl=true&os-OS_X_gl=true&os-OS_X_metal=true&os-Windows_dx=true&os-Windows_dx12=true&os-Windows_gl=true&os-Windows_vulkan=true&pu-dGPU=true&pu-iGPU=true&pu-GPU=true&arch-ARM=true&arch-unknown=true&arch-x86=true&base=device

I think that would be a possible solution but we have to be careful that the resolutions the tests have been rendered on are consistent. In the test you mentioned the GTX 1660 TI reports higher framerates than the RTX 2080 TI very likely due to the resolution.

<img width="725" alt="Screenshot 2020-10-02 at 15 32 32" src="https://user-images.githubusercontent.com/11870513/94929104-eeb6f180-04c4-11eb-99d8-e8079c6e9f82.png">

puckey commented:
Good point – on desktop machines we could filter by '1920x1080' resolutions. But not sure what to do with mobile – we could perhaps even match up screen size to the device? hm

TimvanScherpenzeel commented:
> Good point – on desktop machines we could filter by '1920x1080' resolutions. But not sure what to do with mobile – we could perhaps even match up screen size to the device? hm

I think on desktop devices I think it makes sense to filter (or come up with some formula to describe fps per pixel) and I wouldn't worry about screen size difference on mobile as I presume they tested on the physical phone and not an external screen so the fps in that sense is "real".

TimvanScherpenzeel commented:
GTX 1660 TI -> 723.1 / (1195 x 768) = 0,000787896617852
RTX 2080 TI -> 656.8 / (1920 x 1080) = 0,00031674382716

Sorting by the lowest would then yield the highest relative performance. We might run into some floating point error issues at some point but generally this would work I think

puckey commented:
> > Good point – on desktop machines we could filter by '1920x1080' resolutions. But not sure what to do with mobile – we could perhaps even match up screen size to the device? hm
> 
> I think on desktop devices it makes sense to filter or come up with some formula to describe fps per pixel (this could raise issues with pixel density). I wouldn't worry about screen size difference on mobile as I presume they tested on the physical phone and not an external screen so the fps in that sense is "real".

A GPU could be used in phones of different screen resolutions. For example the A12 bionic is used in these devices:
- Apple iPad Air 2019
- Apple iPad Mini 5
- Apple iPhone XR
- Apple iPhone XS
- Apple iPhone Xs Max
- Apple iPad Air 2019
- Apple iPad Mini 5
- Apple iPhone XR
- Apple iPhone XS
- Apple iPhone Xs Max

puckey commented:
How about the data structure I just pushed: https://github.com/TimvanScherpenzeel/detect-gpu/pull/29/commits/6056a32874a810701be692f474f70d1bd14108cd

We could use this to both match the performance impact of large displays on desktop as well as better match specific mobile devices.

TimvanScherpenzeel commented:
Great! I think that would work well 👍 

What would the `mobileTiers` and `desktopTiers` now signify?

TimvanScherpenzeel commented on rollup/config.lib.ts:
```diff
@@ -0,0 +1,63 @@
+const name = 'DetectGPU';
+
+// tslint:disable-next-line:typedef
+const plugins = ({ isUMD = false, isCJS = false, isES = false }) => [
```
No need for isUMD anymore

TimvanScherpenzeel commented:
> (and if there is not much of a speed difference, I guess we should go for the smallest footprint gzip size wise)

I'll create a new ticket once this is merged, `leven` should do the trick for now.

TimvanScherpenzeel commented on tsconfig.json:
```diff
@@ -1,9 +1,9 @@
-    "target": "es6",
-    "module": "es2015",
-    "lib": ["es2015", "es2016", "es2017", "dom"],
+    "target": "esnext",
```
Would we still be able to support IE 11? The goal is that everywhere WebGL is supported the library should be able to run.

puckey replied:
> Great! I think that would work well 👍
> 
> What would the `mobileTiers` and `desktopTiers` now signify?

Right now it is using the same fps values as in the test it pulls the data from.

puckey replied:
Yeah, it would be good to have a minified polyfilled version and a bare version, I guess.. That way when you are consuming it for use in a web app  using webpack for example, it can use its own browserlist / polyfill setup etc.

puckey replied:
I have been looking into this more, and while I was expecting rollup to offer a backwards compatible option to load chunks using a combination of Fetch/eval it does not seem to do so.. Another option would be to move the `json` files to a public directory and load them using fetch/xmlrequest

puckey commented:
More work on this today – amongst other things:
- instead of using chunks, this version loads json files using [unfetch](https://github.com/developit/unfetch)
- `cjs` and `amd` versions are typescript compiled as `es6` / `es2015`
- the mobile data includes model names for testing purposes, which we should turn off again for production to limit size
- limited benchmark data to Linux_gl / Windows_gl / OS_X_gl / iOS_metal / Android_gl
- added in untested refactor of deobfuscateAppleGpu (with some small byte size changed)
- inlined `cleanRendererString` and renamed it to `getGPUVersion`
- the bare amd version is now 2.13kb gzipped
- For tests it has a `loadBenchmarks` option which can be used to load files in other ways (like `import()` in the tests case)

todo:
- [ ] remove `internal` types exported to `dist/internal`
- [ ] rename `getSupportedWebGLContext` to `getWebGLContext`
- [ ] finish off deobfuscateAppleGpu
- [ ] test on various browsers / devices

puckey commented:
I think functionality wise things are nearly done with the refactor. I spent some time today looking into the deobfuscation of the apple gpu's. I refactored `deobfuscateAppleGpu` to return multiple possible renderers and refactored getGPUTier to be able to work with an array of renderers internally.

I did some brief tests on browserstack and it looks like the library is now successfully detecting most apple devices.

puckey commented:
I also clarified the api a bit by moving overridable parameters together:

```typescript
type GetGPUTierParameters = {
  glContext?: WebGLRenderingContext | WebGL2RenderingContext;
  failIfMajorPerformanceCaveat?: boolean;
  mobileTiers?: number[];
  desktopTiers?: number[];
  benchmarksUrl?: string;
  override?: {
    renderer?: string;
    isIPad?: boolean;
    isMobile?: boolean;
    screen?: { width: number; height: number };
    loadBenchmarks?: (file: string) => Promise<ModelEntry[] | undefined>;
  };
}

TimvanScherpenzeel commented:
Great! I noticed there are quite of few linting errors, would you like me to directly fix it on your branch?

TimvanScherpenzeel commented:
> I also clarified the api a bit by moving overridable parameters together:
> 
> ```ts
> type GetGPUTierParameters = {
>   glContext?: WebGLRenderingContext | WebGL2RenderingContext;
>   failIfMajorPerformanceCaveat?: boolean;
>   mobileTiers?: number[];
>   desktopTiers?: number[];
>   benchmarksUrl?: string;
>   override?: {
>     renderer?: string;
>     isIPad?: boolean;
>     isMobile?: boolean;
>     screen?: { width: number; height: number };
>     loadBenchmarks?: (file: string) => Promise<ModelEntry[] | undefined>;
>   };
> }
> ```

Looks good, I like the idea.

TimvanScherpenzeel commented:
> Great! I noticed there are quite of few linting errors, would you like me to directly fix it on your branch?

I have about 13 large-ish commits on stand-by fixing all lint errors and improving the readability. Please let me know if I can push this to your branch.

Regarding the `dist/` directory, I think we should remove it from git, this will make future PR's also a lot easier to handle.

Notes for me:
- [ ]  Update README / documentation
- [ ]  Create CHANGELOG informing users with breaking changes
- [ ]  Do some more extensive testing
- [ ]  Inform previous contributors, see if they have feedback
- [ ]  Clean up, prepare for a 2.0 alpha release

Overall I am really happy with all the new additions (especially the more accurate reporting and that the work of #26 is included in here). Thank you for all the effort so far 👍 !

puckey commented:
> > Great! I noticed there are quite of few linting errors, would you like me to directly fix it on your branch?
> 
> I have about 13 large-ish commits on stand-by fixing all lint errors and improving the readability. Please let me know if I can push this to your branch.

Yes, go ahead. How about we merge the pull request into the development branch and you take it from there?

After checking with TSLint I was wondering about some of the rules:
- `expected arrow-call-signature to have a typedef`:  in my opinion adding return types to arrow functions is unnecessary unless typescript is unable to infer. It also leads to more work during development, since you keep having to adjust every single arrow function type whenever a large change is made up the chain. I would suggest we lean more on typescript's power to infer types (it will warn when any type becomes untyped any anyway, since we are in strict mode)
- `no-shadowed-variable`: in my experience this can lead to less legible code when followed strictly.. 
- `prefer-const`: this rule leads to situations where it doesn't want you destructuring into a let when you are only reassigning one of the destructured values - which seems a bit pedantic to me..

Another option would be to skip eslint altogether and opt for a combination of prettier & typescript to cover most code quality cases...

> Regarding the `dist/` directory, I think we should remove it from git, this will make future PR's also a lot easier to handle.

Good point!

TimvanScherpenzeel commented:
> > > Great! I noticed there are quite of few linting errors, would you like me to directly fix it on your branch?
> > 
> > 
> > I have about 13 large-ish commits on stand-by fixing all lint errors and improving the readability. Please let me know if I can push this to your branch.
> 
> Yes, go ahead. How about we merge the pull request into the development branch and you take it from there?
> 
> After checking with TSLint I was wondering about some of the rules:
> 
> * `expected arrow-call-signature to have a typedef`:  in my opinion adding return types to arrow functions is unnecessary unless typescript is unable to infer. It also leads to more work during development, since you keep having to adjust every single arrow function type whenever a large change is made up the chain. I would suggest we lean more on typescript's power to infer types (it will warn when any type becomes untyped any anyway, since we are in strict mode)
> * `no-shadowed-variable`: in my experience this can lead to less legible code when followed strictly..
> * `prefer-const`: this rule leads to situations where it doesn't want you destructuring into a let when you are only reassigning one of the destructured values - which seems a bit pedantic to me..
> 
> Another option would be to skip eslint altogether and opt for a combination of prettier & typescript to cover most code quality cases...
> 
> > Regarding the `dist/` directory, I think we should remove it from git, this will make future PR's also a lot easier to handle.
> 
> Good point!

Sounds good! I'll merge it into development.
